samples/subsys/modbus/tcp_gateway/src/main.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2020 PHYTEC Messtechnik GmbH
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/util.h>
+#include <modbus/modbus.h>
+#include <net/socket.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tcp_gateway, LOG_LEVEL_INF);
+
+#define MODBUS_TCP_PORT 502
+
+static struct modbus_adu tmp_adu;
+static int backend;
+
+const static struct modbus_iface_param backend_param = {
+	.mode = MODBUS_MODE_RTU,
+	.rx_timeout = 50000,
+	.serial = {
+		.baud = 19200,
+		.parity = UART_CFG_PARITY_NONE,
+	},
+};
+
+static int init_backend_iface(void)
+{
+	const char bend_name[] = {DT_PROP(DT_INST(0, modbus_rtu), label)};
+
+	backend = modbus_iface_get_by_name(bend_name);
+	if (backend < 0) {
+		LOG_ERR("Failed to get iface index for %s",
+			log_strdup(bend_name));
+		return -ENODEV;
+	}
+
+	return modbus_init_client(backend, backend_param);
+}
+
+static int modbus_tcp_reply(int client, struct modbus_adu *adu)
+{
+	uint8_t header[MODBUS_MBAP_AND_FC_LENGTH];
+
+	modbus_raw_put_header(adu, header);
+	if (send(client, header, sizeof(header), 0) < 0) {
+		return errno;
+	}
+
+	if (send(client, adu->data, adu->length, 0) < 0) {
+		return errno;
+	}
+
+	return 0;
+}
+
+static int modbus_tcp_connection(int client)
+{
+	uint8_t header[MODBUS_MBAP_AND_FC_LENGTH];
+	int rc;
+	int data_len;
+
+	rc = recv(client, header, sizeof(header), MSG_WAITALL);
+	if (rc <= 0) {
+		return errno;
```

Comment:
nit: I'd return `-errno` so that a negative value would indicate an error in all cases. Same in other places returning errno.